README
This is a xserver-xorg input driver for devices supporting
Linux Multi-Touch protocol.

Author: Mika Kuoppala <mika.kuoppala at-here nokia.com>

This driver is based on apple touchpad driver by Henrik Rydberg

Installation howto:

You need to add correct input class definitio to xorg.conf
Look at conf/xorg.conf.section and cut paste it into your xorg.conf

The kernel device driver needs to pass correct MultiTouch protocol events to
this driver. Also note that BTN_TOUCH needs to be provided by the kernel
driver for udev to correctly set MatchIsTouchscreen capability in xserver.





debian/changelog
xserver-xorg-input-mtev (0.1.13) unstable; urgency=low
   * Adapted to use libmtdev, supporting kernel protocol A and B

 -- James Ketrenos <jketreno@linux.intel.com> Mon, 6 Dec 2010 14:56:03 -0800

xserver-xorg-input-mtev (0.1.12) unstable; urgency=low

  * Send motion events before button down

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Wed, 27 Oct 2010 13:40:02 +0300

xserver-xorg-input-mtev (0.1.11) unstable; urgency=low

  * Added orientation config options: SwapAxes, InvertX, InvertY.

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Tue, 26 Oct 2010 13:57:06 +0300

xserver-xorg-input-mtev (0.1.10) unstable; urgency=low

  * Dummy version bump to get messed up and released tags in sync :(

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Wed, 15 Sep 2010 14:22:53 +0300

xserver-xorg-input-mtev (0.1.9) unstable; urgency=low

  * Removed depracated xallocs

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Wed, 15 Sep 2010 14:17:58 +0300

xserver-xorg-input-mtev (0.1.8) unstable; urgency=low

  * Fixes: NB#167575 If kernel driver does't expose minor width,
  assume it is the same as major width. As stated by Multi-Touch
  protocol

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Wed, 19 May 2010 15:38:43 +0300

xserver-xorg-input-mtev (0.1.7) unstable; urgency=low

  * correct patchlevel for the module too

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Tue, 20 Apr 2010 14:37:04 +0300

xserver-xorg-input-mtev (0.1.6) unstable; urgency=low

  * fixed: buffer size was never grown due to redefinition

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Fri, 09 Apr 2010 13:23:18 +0300

xserver-xorg-input-mtev (0.1.5) unstable; urgency=low

  * bigger input buffer size

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Wed, 31 Mar 2010 15:47:03 +0300

xserver-xorg-input-mtev (0.1.4) unstable; urgency=low

  * Fix valuator overflow with >6 fingers

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Tue, 30 Mar 2010 11:58:08 +0300

xserver-xorg-input-mtev (0.1.3) unstable; urgency=low

  * Fix double free on uninit

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Fri, 26 Mar 2010 14:55:02 +0200

xserver-xorg-input-mtev (0.1.2) unstable; urgency=low

  * cleanup

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Thu, 04 Mar 2010 18:32:05 +0200

xserver-xorg-input-mtev (0.1.1) unstable; urgency=low

  * clamp fingers to six

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Tue, 16 Feb 2010 16:59:00 +0200

xserver-xorg-input-mtev (0.1) unstable; urgency=low

  * Initial version. Based heavily on a work by Hendrik Rydberg (multitouch)

 -- Mika Kuoppala <mika.kuoppala@nokia.com>  Wed, 16 Dec 2009 17:21:33 +0200


src/hw.c
/***************************************************************************
 *
 * Multitouch protocol X driver
 * Copyright (C) 2008 Henrik Rydberg <rydberg@euromail.se>
 * Copyright (C) 2009,2010 Nokia Corporation
 *
 * Adaptation to libmtdev and Linux multi-touch-protocol B (slotted)
 * Copyright (C) 2010 Intel Corporation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 **************************************************************************/

#include <string.h>
#include <linux/input.h>
#include <xf86.h>
#include <mtdev.h>

#include "hw.h"

void hw_init(struct mtev_hw_state *hw)
{
	int i;
	memset(hw, 0, sizeof(struct mtev_hw_state));
	for (i = 0; i < HW_MAX_SLOTS; i++)
	    hw->slot[i].tracking_id = INVALID_TRACKING_ID;
}

bool hw_read(struct mtev_hw_state *hw, const struct input_event* ev)
{
	// xf86Msg(X_INFO, "event: %d %d %d\n", ev->type, ev->code, ev->value);

	switch (ev->type) {
	case EV_SYN:
		switch (ev->code) {
		case SYN_REPORT:
			return 1;

		case SYN_MT_REPORT:
			xf86Msg(X_ERROR, "libmtdev sent SYN_MT_REPORT");
			break;
		}
		break;

	case EV_ABS:
		if (ev->code == ABS_MT_SLOT) {
			if (ev->value >= HW_MAX_SLOTS) {
				xf86Msg(X_ERROR, "Slot usage (%d) exceeds limit of %d", ev->value, HW_MAX_SLOTS);
				hw->current_slot = INVALID_SLOT;
			} else
				hw->current_slot = ev->value;
			break;
		}

		if (hw->current_slot == INVALID_SLOT)
			break;

		switch (ev->code) {
		case ABS_MT_POSITION_X:
			hw->slot[hw->current_slot].position_x = ev->value;
			break;
		case ABS_MT_POSITION_Y:
			hw->slot[hw->current_slot].position_y = ev->value;
			break;
		case ABS_MT_TOUCH_MAJOR:
			hw->slot[hw->current_slot].touch_major = ev->value;
			break;
		case ABS_MT_TOUCH_MINOR:
			hw->slot[hw->current_slot].touch_minor = ev->value;
			break;
		case ABS_MT_WIDTH_MAJOR:
			hw->slot[hw->current_slot].width_major = ev->value;
			break;
		case ABS_MT_WIDTH_MINOR:
			hw->slot[hw->current_slot].width_minor = ev->value;
			break;
		case ABS_MT_ORIENTATION:
			hw->slot[hw->current_slot].orientation = ev->value;
			break;
		case ABS_MT_PRESSURE:
			hw->slot[hw->current_slot].pressure = ev->value;
			break;
		case ABS_MT_TRACKING_ID:
			if (ev->value == -1) /* Slot contact has been released */
				hw->slot[hw->current_slot].tracking_id = INVALID_TRACKING_ID;
			else
				hw->slot[hw->current_slot].tracking_id = hw->current_slot;//ev->value;
			break;
		}
	}

	return 0;
}


src/hw.h
/***************************************************************************
 *
 * Multitouch protocol X driver
 * Copyright (C) 2008 Henrik Rydberg <rydberg@euromail.se>
 * Copyright (C) 2009,2010 Nokia Corporation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 **************************************************************************/

#ifndef HWDATA_H
#define HWDATA_H

#include "common.h"

struct input_event;

// Max touch points we gonna get out from kernel layer
#define HW_MAX_SLOTS 10
#define INVALID_SLOT -1
#define INVALID_TRACKING_ID -1

struct mtev_touch_point {
	int touch_major, touch_minor;
	int width_major, width_minor;
	int orientation;
	int position_x, position_y;
	int pressure;
	int tracking_id;
};

struct mtev_hw_state {
	struct mtev_touch_point slot[HW_MAX_SLOTS];
	int current_slot;
};

void hw_init(struct mtev_hw_state *hw);
bool hw_read(struct mtev_hw_state *hw, const struct input_event* ev);

#endif


src/caps.c
/***************************************************************************
 *
 * Multitouch protocol X driver
 * Copyright (C) 2008 Henrik Rydberg <rydberg@euromail.se>
 * Copyright (C) 2009,2010 Nokia Corporation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 **************************************************************************/

#include "caps.h"
#include <errno.h>
#include <string.h>
#include <xf86.h>
#include <xf86Xinput.h>

#define SETABS(c, x, map, key, fd)					\
	c->has_##x = getbit(map, key) && getabs(&c->abs_##x, key, fd)

#define ADDCAP(s, c, x) strcat(s, c->has_##x ? " " #x : "")

static const int bits_per_long = 8 * sizeof(long);

static inline int nlongs(int nbit)
{
	return (nbit + bits_per_long - 1) / bits_per_long;
}

static inline bool getbit(const unsigned long* map, int key)
{
	return (map[key / bits_per_long] >> (key % bits_per_long)) & 0x01;
}

static bool getabs(struct input_absinfo *abs, int key, int fd)
{
	int rc;
	SYSCALL(rc = ioctl(fd, EVIOCGABS(key), abs));
	return rc >= 0;
}

int caps_read(struct mtev_caps *caps, int fd)
{
	unsigned long evbits[nlongs(EV_MAX)];
	unsigned long absbits[nlongs(ABS_MAX)];
	unsigned long keybits[nlongs(KEY_MAX)];
	int rc;

	memset(caps, 0, sizeof(struct mtev_caps));

	SYSCALL(rc = ioctl(fd, EVIOCGBIT(EV_SYN, sizeof(evbits)), evbits));
	if (rc < 0) {
		xf86Msg(X_ERROR, "mtev: EV_SYN needed but missing\n");
		return rc;
	}

	SYSCALL(rc = ioctl(fd, EVIOCGBIT(EV_KEY, sizeof(keybits)), keybits));
	if (rc < 0) {
		xf86Msg(X_ERROR, "mtev: EV_KEY needed but missing\n");
		return rc;
	}
	SYSCALL(rc = ioctl(fd, EVIOCGBIT(EV_ABS, sizeof(absbits)), absbits));
	if (rc < 0) {
		xf86Msg(X_ERROR, "mtev: EV_ABS needed but missing\n");
		return rc;
	}

	caps->has_left = getbit(keybits, BTN_LEFT);
	caps->has_middle = getbit(keybits, BTN_MIDDLE);
	caps->has_right = getbit(keybits, BTN_RIGHT);

	SETABS(caps, touch_major, absbits, ABS_MT_TOUCH_MAJOR, fd);
	SETABS(caps, touch_minor, absbits, ABS_MT_TOUCH_MINOR, fd);
	SETABS(caps, width_major, absbits, ABS_MT_WIDTH_MAJOR, fd);
	SETABS(caps, width_minor, absbits, ABS_MT_WIDTH_MINOR, fd);
	SETABS(caps, orientation, absbits, ABS_MT_ORIENTATION, fd);
	SETABS(caps, position_x, absbits, ABS_MT_POSITION_X, fd);
	SETABS(caps, position_y, absbits, ABS_MT_POSITION_Y, fd);
	SETABS(caps, tracking_id, absbits, ABS_MT_TRACKING_ID, fd);

	caps->has_mtdata = caps->has_position_x && caps->has_position_y;

	return 0;
}

void caps_output(const struct mtev_caps *caps)
{
	char line[1024];
	memset(line, 0, sizeof(line));
	ADDCAP(line, caps, left);
	ADDCAP(line, caps, middle);
	ADDCAP(line, caps, right);
	ADDCAP(line, caps, mtdata);
	ADDCAP(line, caps, touch_major);
	ADDCAP(line, caps, touch_minor);
	ADDCAP(line, caps, width_major);
	ADDCAP(line, caps, width_minor);
	ADDCAP(line, caps, orientation);
	ADDCAP(line, caps, tracking_id);
	ADDCAP(line, caps, position_x);
	ADDCAP(line, caps, position_y);

	xf86Msg(X_INFO, "mtev: caps:%s\n", line);
	if (caps->has_touch_major)
		xf86Msg(X_INFO, "mtev: touch major: %d %d\n",
			caps->abs_touch_major.minimum,
			caps->abs_touch_major.maximum);
	if (caps->has_touch_minor)
		xf86Msg(X_INFO, "mtev: touch minor: %d %d\n",
			caps->abs_touch_minor.minimum,
			caps->abs_touch_minor.maximum);
	if (caps->has_width_major)
		xf86Msg(X_INFO, "mtev: width: %d %d\n",
			caps->abs_width_major.minimum,
			caps->abs_width_major.maximum);
	if (caps->has_orientation)
		xf86Msg(X_INFO, "mtev: orientation: %d %d\n",
			caps->abs_orientation.minimum,
			caps->abs_orientation.maximum);
	if (caps->has_tracking_id)
		xf86Msg(X_INFO, "mtev: tracking_id: %d %d\n",
			caps->abs_tracking_id.minimum,
			caps->abs_tracking_id.maximum);
	if (caps->has_position_x)
		xf86Msg(X_INFO, "mtev: position_x: %d %d\n",
			caps->abs_position_x.minimum,
			caps->abs_position_x.maximum);
	if (caps->has_position_y)
		xf86Msg(X_INFO, "mtev: position_y: %d %d\n",
			caps->abs_position_y.minimum,
			caps->abs_position_y.maximum);
}


src/caps.h
/***************************************************************************
 *
 * Multitouch protocol X driver
 * Copyright (C) 2008 Henrik Rydberg <rydberg@euromail.se>
 * Copyright (C) 2009,2010 Nokia Corporation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 **************************************************************************/

#ifndef CAPABILITIES_H
#define CAPABILITIES_H

#include "common.h"
#include <linux/input.h>

struct mtev_caps {
	bool has_left, has_middle;
	bool has_right, has_mtdata;
	bool has_touch_major, has_touch_minor;
	bool has_width_major, has_width_minor;
	bool has_orientation, has_tracking_id;
	bool has_position_x, has_position_y;
	struct input_absinfo abs_touch_major;
	struct input_absinfo abs_touch_minor;
	struct input_absinfo abs_width_major;
	struct input_absinfo abs_width_minor;
	struct input_absinfo abs_orientation;
	struct input_absinfo abs_position_x;
	struct input_absinfo abs_position_y;
	struct input_absinfo abs_tracking_id;
};

int caps_read(struct mtev_caps *caps, int fd);
void caps_output(const struct mtev_caps *caps);

#endif


src/multitouch.c
/***************************************************************************
 *
 * Multitouch X driver
 * Copyright (C) 2008 Henrik Rydberg <rydberg@euromail.se>
 * Copyright (C) 2009,2010 Nokia Corporation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 **************************************************************************/

#define MODULEVENDORSTRING "Nokia"
/* #define NO_RIGHT_MOUSE_BUTTON_EMULATION */

#include "xorg-server.h"
#include <xorg/exevents.h>
#include <xorg/xserver-properties.h>
#include <X11/Xatom.h>
#include <xf86.h>
#include <xf86_OSproc.h>
#include <xf86Xinput.h>

#include "common.h"
#include "mtouch.h"

static const char* const axis_labels_str[] = {
	AXIS_LABEL_PROP_ABS_MT_POSITION_X,
	AXIS_LABEL_PROP_ABS_MT_POSITION_Y,
	AXIS_LABEL_PROP_ABS_MT_TOUCH_MAJOR,
	AXIS_LABEL_PROP_ABS_MT_TOUCH_MINOR,
	AXIS_LABEL_PROP_ABS_MT_TRACKING_ID,
};

static void pointer_control(DeviceIntPtr dev, PtrCtrl *ctrl)
{
	xf86Msg(X_INFO, "pointer_control\n");
}

static int pointer_property(DeviceIntPtr dev,
			    Atom property,
			    XIPropertyValuePtr prop,
			    BOOL checkonly)
{
	xf86Msg(X_INFO, "pointer_property\n");
	return Success;
}

static void init_axes_labels(Atom* labels, int num_labels)
{
	int i;

	for (i = 0 ; i < num_labels; i++) {
		labels[i] = MakeAtom(axis_labels_str[i % MT_AXIS_PER_FINGER],
				     strlen(axis_labels_str[i % MT_AXIS_PER_FINGER]),
				     TRUE);
	}
}

static int init_properties(DeviceIntPtr dev)
{
	static const char* const strMaxContacts = "Max Contacts";
	static const char* const strAxesPerContact = "Axes Per Contact";
	int rc;

	Atom labelMaxContacts;
	Atom labelAxesPerContact;

	int max_contacts = MT_NUM_FINGERS;
	int axes_per_contact = MT_AXIS_PER_FINGER;

	labelMaxContacts = MakeAtom(strMaxContacts,
				    strlen(strMaxContacts), TRUE);
	labelAxesPerContact = MakeAtom(strAxesPerContact,
				       strlen(strAxesPerContact), TRUE);

	rc = XIChangeDeviceProperty(dev,
				    labelMaxContacts,
				    XA_INTEGER,
				    8,
				    PropModeReplace,
				    1,
				    &max_contacts,
				    TRUE);
	if (rc != Success)
		return rc;

	XISetDevicePropertyDeletable(dev, labelMaxContacts, FALSE);


	rc = XIChangeDeviceProperty(dev,
				    labelAxesPerContact,
				    XA_INTEGER,
				    8,
				    PropModeReplace,
				    1,
				    &axes_per_contact,
				    TRUE);

	if (rc != Success)
		return rc;

	XISetDevicePropertyDeletable(dev, labelAxesPerContact, FALSE);

	return Success;
}

static int device_init(DeviceIntPtr dev, InputInfoPtr pInfo)
{
	struct mtev_mtouch *mt = pInfo->private;
	Atom atom;
	int i;
	int j;
	unsigned char map[MT_NUM_BUTTONS + 1];
	Atom btn_labels[MT_NUM_BUTTONS] = { 0 };
	Atom axes_labels[MT_NUM_VALUATORS] = { 0, };
	int r;

	if (MT_NUM_VALUATORS > MAX_VALUATORS) {
		xf86Msg(X_ERROR, "MT_NUM_VALUATORS(%d) > MAX_VALUATORS(%d)\n",
			MT_NUM_VALUATORS, MAX_VALUATORS);
		return BadValue;
	}

	for (i = 0; i < MT_NUM_BUTTONS; i++)
		btn_labels[i] = XIGetKnownProperty(BTN_LABEL_PROP_BTN_UNKNOWN);

	atom = XIGetKnownProperty(BTN_LABEL_PROP_BTN_LEFT);
	btn_labels[0] = atom;

	init_axes_labels(axes_labels, MT_NUM_VALUATORS);

	r = init_properties(dev);
	if (r != Success)
		return r;

	pInfo->fd = xf86OpenSerial(pInfo->options);
	if (pInfo->fd < 0) {
		xf86Msg(X_ERROR, "mtev: cannot open device\n");
		return !Success;
	}
	if (mtouch_configure(mt, pInfo->fd)) {
		xf86Msg(X_ERROR, "mtev: cannot configure device\n");
		return !Success;
	}
	xf86CloseSerial(pInfo->fd);

	for (i = 0; i < MT_NUM_BUTTONS+1; i++)
		map[i] = i;

	InitPointerDeviceStruct((DevicePtr)dev,
				map,
				MT_NUM_BUTTONS,
				btn_labels,
				pointer_control,
				GetMotionHistorySize(),
				MT_NUM_VALUATORS,
				axes_labels);

	for (i = 0; i < MT_NUM_FINGERS; i++) {
		for (j = 0; j < MT_AXIS_PER_FINGER; j++) {
			const int val = (i * MT_AXIS_PER_FINGER) + j;
			int min;
			int max;

			switch (j) {
			case 0:
				min = mt->caps.abs_position_x.minimum;
				max = mt->caps.abs_position_x.maximum;
				break;
			case 1:
				min = mt->caps.abs_position_y.minimum;
				max = mt->caps.abs_position_y.maximum;
				break;
			case 2:
				min = mt->caps.abs_touch_major.minimum;
				max = mt->caps.abs_touch_major.maximum;
				break;
			case 3:
				if (mt->caps.has_touch_minor) {
					min = mt->caps.abs_touch_minor.minimum;
					max = mt->caps.abs_touch_minor.maximum;
				} else {
					min = mt->caps.abs_touch_major.minimum;
					max = mt->caps.abs_touch_major.maximum;
				}
				break;
			case 4: // Tracking id
				min = mt->caps.abs_tracking_id.minimum;
				max = mt->caps.abs_tracking_id.maximum >
					(MT_NUM_FINGERS-1) ?
					(MT_NUM_FINGERS-1) :
					mt->caps.abs_tracking_id.maximum;
				break;
			default:
				return BadValue;
			}

			xf86InitValuatorAxisStruct(dev, val, axes_labels[val],
						   min,
						   max,
						   1, 0, 1, Absolute);
			xf86InitValuatorDefaults(dev, val);
		}
	}

	XIRegisterPropertyHandler(dev, pointer_property, NULL, NULL);

	return Success;
}

static int device_on(InputInfoPtr pInfo)
{
	struct mtev_mtouch *mt = pInfo->private;
	pInfo->fd = xf86OpenSerial(pInfo->options);
	if (pInfo->fd < 0) {
		xf86Msg(X_ERROR, "mtev: cannot open device\n");
		return !Success;
	}
	if (mtouch_open(mt, pInfo->fd)) {
		xf86Msg(X_ERROR, "mtev: cannot grab device\n");
		return !Success;
	}
	xf86AddEnabledDevice(pInfo);
	return Success;
}

static int device_off(InputInfoPtr pInfo)
{
	struct mtev_mtouch *mt = pInfo->private;
	xf86RemoveEnabledDevice(pInfo);
	if(mtouch_close(mt, pInfo->fd)) {
		xf86Msg(X_WARNING, "mtev: cannot ungrab device\n");
	}
	xf86CloseSerial(pInfo->fd);
	return Success;
}

static int device_close(InputInfoPtr pInfo)
{
	return Success;
}

#define BUTTON_1 (1L<<1)
#define BUTTON_2 (1L<<2)
#define BUTTON_3 (1L<<3)

static void process_state(InputInfoPtr pInfo,
			  const struct mtev_mtouch *mt)
{

	const struct mtev_touch_point *tp;
	static int pdown = 0;
#ifndef NO_RIGHT_MOUSE_BUTTON_EMULATION
	static int buttons = 0;
#endif
	int valuators[MAX_VALUATORS];
	int down;
	int valix;
	int contacts;

	contacts = valix = down = 0;

	while ((tp = mtouch_get_contact(mt, contacts)) != NULL) {
		contacts++;

		// We don't do remapping of tracking id's so
		// make sure clients don't see too high tracking_id numbers
		if (tp->tracking_id < MT_NUM_FINGERS) {
			int x;
			int y;

			x = tp->position_x;
			y = tp->position_y;

			if (mt->invert_x)
				x = mt->caps.abs_position_x.maximum - x +
					mt->caps.abs_position_x.minimum;

			if (mt->invert_y)
				y = mt->caps.abs_position_y.maximum - y
					+ mt->caps.abs_position_y.minimum;

			if (mt->swap_xy) {
				const int tmp = y;
				y = x;
				x = tmp;
			}

                        switch (mt->rotation)
			{
				case 1:
				{
					const int tmp = (mt->caps.abs_position_x.maximum - x) * (mt->caps.abs_position_y.maximum - mt->caps.abs_position_x.minimum) / (mt->caps.abs_position_x.maximum - mt->caps.abs_position_x.minimum) + mt->caps.abs_position_y.minimum;
                                	x = (y - mt->caps.abs_position_y.minimum) * (mt->caps.abs_position_x.maximum - mt->caps.abs_position_x.minimum) / (mt->caps.abs_position_y.maximum - mt->caps.abs_position_y.minimum) + mt->caps.abs_position_x.minimum;
					y = tmp;
					break;
				}
				case 2:
				{
					x = mt->caps.abs_position_x.maximum - x + mt->caps.abs_position_x.minimum;
					y = mt->caps.abs_position_y.maximum - y + mt->caps.abs_position_y.minimum;
					break;
				}
				case 3:
				{
					const int tmp = (x - mt->caps.abs_position_x.minimum) * (mt->caps.abs_position_y.maximum - mt->caps.abs_position_y.minimum) / (mt->caps.abs_position_x.maximum - mt->caps.abs_position_x.minimum) + mt->caps.abs_position_y.minimum;
                                        x = (mt->caps.abs_position_y.maximum - y) * (mt->caps.abs_position_x.maximum - mt->caps.abs_position_x.minimum) / (mt->caps.abs_position_y.maximum - mt->caps.abs_position_y.minimum) + mt->caps.abs_position_x.minimum;
                                        y = tmp;
					break;
				}
                        }
			valuators[valix++] = x;
			valuators[valix++] = y;
			valuators[valix++] = tp->touch_major;

			if (mt->caps.has_touch_minor)
				valuators[valix++] = tp->touch_minor;
			else
				valuators[valix++] = tp->touch_major;

			valuators[valix++] = tp->tracking_id;

			down++;
		}

		// Don't deliver more than MaxContacts
		if (down >= MT_NUM_FINGERS)
			break;
	}

	/* Some x-clients assume they get motion events before button down */
	if (down)
		xf86PostMotionEventP(pInfo->dev, TRUE,
				     0, down * MT_AXIS_PER_FINGER, valuators);

#ifdef NO_RIGHT_MOUSE_BUTTON_EMULATION
	if(down && pdown == 0)
		xf86PostButtonEventP(pInfo->dev, TRUE,
				     1, 1,
				     0, down * MT_AXIS_PER_FINGER, valuators);
	else if (down == 0 && pdown)
		xf86PostButtonEvent(pInfo->dev, TRUE, 1, 0, 0, 0);

	pdown = !!down;
#else // Right Mouse Button Emulation
	if(down > pdown) {
		// Button Down
		if(down == 1 && !(buttons & BUTTON_1)) {
			buttons = BUTTON_1;
			xf86PostButtonEventP(pInfo->dev, TRUE,
					     1, 1,
					     0, down * MT_AXIS_PER_FINGER, valuators);
		} else if (down == 2 && !(buttons & BUTTON_3)) {
			buttons |= BUTTON_3;
			xf86PostButtonEventP(pInfo->dev, TRUE,
					     3, 1,
					     0, down * MT_AXIS_PER_FINGER, valuators);
		}
	} else if (down < pdown && down < 2) {
		if(buttons & BUTTON_3) {
			xf86PostButtonEventP(pInfo->dev, TRUE,
					     3, 0,
					     0, down * MT_AXIS_PER_FINGER, valuators);
			buttons &= ~BUTTON_3;
		}
		/* if down == 0, left button handled below */
	}

	if(down == 0) {
		if(buttons & BUTTON_3)
			xf86PostButtonEventP(pInfo->dev, TRUE,
					     3, 0,
					     0, down * MT_AXIS_PER_FINGER, valuators);
		if(buttons & BUTTON_1)
			xf86PostButtonEvent(pInfo->dev, TRUE, 1, 0, 0, 0);
		buttons = 0;
	}

	pdown = down;
#endif // NO_RIGHT_MOUSE_BUTTON_EMULATION
}

/* called for each full received packet from the touchpad */
static void read_input(InputInfoPtr pInfo)
{
	struct mtev_mtouch *mt = pInfo->private;
	while (mtouch_read_synchronized_event(mt, pInfo->fd)) {
		process_state(pInfo, mt);
	}
}

static Bool device_control(DeviceIntPtr dev, int mode)
{
	InputInfoPtr pInfo = dev->public.devicePrivate;
	switch (mode) {
	case DEVICE_INIT:
		xf86Msg(X_INFO, "device control: init\n");
		return device_init(dev, pInfo);
	case DEVICE_ON:
		xf86Msg(X_INFO, "device control: on\n");
		return device_on(pInfo);
	case DEVICE_OFF:
		xf86Msg(X_INFO, "device control: off\n");
		return device_off(pInfo);
	case DEVICE_CLOSE:
		xf86Msg(X_INFO, "device control: close\n");
		return device_close(pInfo);
	default:
		xf86Msg(X_INFO, "device control: default\n");
		return BadValue;
	}
}

static int preinit(InputDriverPtr drv, InputInfoPtr pInfo, int flags)
{
	struct mtev_mtouch *mt;
	if (!pInfo)
		goto error;
	mt = calloc(1, sizeof(struct mtev_mtouch));
	if (!mt)
		goto error;

	pInfo->type_name = XI_TOUCHSCREEN;
	pInfo->device_control = device_control;
	pInfo->read_input = read_input;
	pInfo->private = mt;
	pInfo->flags = 0;

	xf86CollectInputOptions(pInfo, NULL);
	//xf86OptionListReport(pInfo->options);
	xf86ProcessCommonOptions(pInfo, pInfo->options);


	mt->swap_xy = xf86SetBoolOption(pInfo->options, "SwapAxes", FALSE);
	mt->invert_x = xf86SetBoolOption(pInfo->options, "InvertX", FALSE);
	mt->invert_y = xf86SetBoolOption(pInfo->options, "InvertY", FALSE);

        // o = 0
        // 1 = CW 90
        // 2 = CW 180
        // 3 = CW 270
        mt->rotation = xf86SetIntOption(pInfo->options, "Rotation", 0);

        if (mt->rotation < 0 || mt->rotation > 3)
               mt->rotation = 0;


	return Success;
error:
	return BadAlloc;
}

static void uninit(InputDriverPtr drv, InputInfoPtr pInfo, int flags)
{
	free(pInfo->private);
	pInfo->private = NULL;
	xf86DeleteInput(pInfo, 0);
}

static InputDriverRec MTEV = {
	.driverVersion = 1,
	.driverName = "mtev",
	.Identify = NULL,
	.PreInit = preinit,
	.UnInit = uninit,
	.module = NULL,
};

static XF86ModuleVersionInfo VERSION = {
	.modname = "mtev",
	.vendor = "Nokia",
	._modinfo1_ = MODINFOSTRING1,
	._modinfo2_ = MODINFOSTRING2,
	.xf86version = XORG_VERSION_CURRENT,
	.majorversion = 0,
	.minorversion = 1,
	.patchlevel = 12,
	.abiclass = ABI_CLASS_XINPUT,
	.abiversion = ABI_XINPUT_VERSION,
	.moduleclass = MOD_CLASS_XINPUT,
	.checksum = {0, 0, 0, 0}
};

static pointer setup(pointer module, pointer options, int *errmaj, int *errmin)
{
	xf86AddInputDriver(&MTEV, module, 0);
	return module;
}

XF86ModuleData mtevModuleData = {
	.vers = &VERSION,
	.setup = &setup,
	.teardown = NULL
};
